### PR TITLE
Execute async option

### LINKED ISSUE #8: Add a task to run PUT file commands
# Problem
Running commands such as `put file:///data.csv @my_csv_stage` is a common (and efficient) way of loading data into a staging area in Snowflake. This doesn't work using the query task since this doesn't seem to be supported with the Snowflake connector for Python, but rather only from SnowSQL. 

# Desired behavior
Investigate and if possible, add a task to enable loading data from files into Snowflake via PUT file command.

# Discussion on Slack
[**View in #prefect-community on Slack**](https://slack.com/archives/CL09KU1K7/p1653385522181629)

![Todd_de_Quincey] **@Todd_de_Quincey:** Prefect 2.0 - Snowflake PUT command from local file error
When attempting to run the below task

```
result = snowflake_query(
        query="put file:///data.csv @my_csv_stage",
        snowflake_credentials=SNOWFLAKE_CREDENTIALS,
)
```

I get the following error

```
11:40:45.794 | ERROR   | Task run 'snowflake_query-eb69c8ef-0' - Encountered exception during execution:
Traceback (most recent call last):
  File "/Users/todddequincey/repos/midnite/data/prefect/.venv/lib/python3.9/site-packages/prefect/engine.py", line 796, in orchestrate_task_run
    result = await task.fn(*args, **kwargs)
  File "/Users/todddequincey/repos/midnite/data/prefect/.venv/lib/python3.9/site-packages/prefect_snowflake/database.py", line 68, in snowflake_query
    response = cursor.execute_async(query, params=params)
  File "/Users/todddequincey/repos/midnite/data/prefect/.venv/lib/python3.9/site-packages/snowflake/connector/cursor.py", line 804, in execute_async
    return self.execute(*args, **kwargs)
  File "/Users/todddequincey/repos/midnite/data/prefect/.venv/lib/python3.9/site-packages/snowflake/connector/cursor.py", line 755, in execute
    sf_file_transfer_agent.execute()
  File "/Users/todddequincey/repos/midnite/data/prefect/.venv/lib/python3.9/site-packages/snowflake/connector/file_transfer_agent.py", line 355, in execute
    self._parse_command()
  File "/Users/todddequincey/repos/midnite/data/prefect/.venv/lib/python3.9/site-packages/snowflake/connector/file_transfer_agent.py", line 852, in _parse_command
    response = self._ret["data"]
KeyError: 'command'
```

Setting a breakpoint at `/Users/todddequincey/repos/midnite/data/prefect/.venv/lib/python3.9/site-packages/snowflake/connector/file_transfer_agent.py::854`  and inspecting the `response = _self_._ret["data"]` object, there is indeed no `command` object in there.

Not sure if this is a user error, but haven’t been able to find much in the way of docs or examples for this.

Note: I have manually checked in snowflake that the user and role has appropriate access to the stage. And inspecting the os path, the file is accessible from the code.

![Anna_Geller] **@Anna_Geller:** Could you share the full flow code showing how you use it in your flow? Hide any credentials

![Todd_de_Quincey] **@Todd_de_Quincey:** That is the full flow at this stage - I stripped everything out to isolate the problem

```
from prefect_snowflake.database import snowflake_query
from prefect import flow

from util.const import AWS_CREDENTIALS, SNOWFLAKE_CREDENTIALS, S3_RAW_BUCKET


@flow
def snowflake_pipeline():
    result = snowflake_query(
        query="put file:///data.csv @my_csv_stage",
        snowflake_credentials=SNOWFLAKE_CREDENTIALS,
    )

    return None


state = snowflake_pipeline()
```

And replacing the query with a select or insert works fine. So I know the connection is fine
Literally just doing a spike to see if Prefect is going to meet our needs

![Anna_Geller] **@Anna_Geller:** A dumb question but is a put command considered a query by Snowflake API?
I think this might be true, based on the top answer [here](https://community.snowflake.com/s/question/0D50Z00009XCitqSAD/how-to-execute-put-file-command-to-load-data-to-snowflake-table-from-pyhon-jupyter-nb), you can't execute put command using a Python connector, you need to do it via SnowSQL
> Snowflake Community

![Todd_de_Quincey] **@Todd_de_Quincey:** It’s listed under their list of SQL commands
<https://docs.snowflake.com/en/sql-reference/sql/put.html>

![Anna_Geller] **@Anna_Geller:** Quote from Snowflake: "You can't execute a PUT with the python connector, you need to use SnowSQL. If you want to execute PUT through python, you'd have to create script file and then execute SnowSQL through an OS command function in python. It's not the cleanest way to go. If you use your own S3 buckets, then you have a lot more options on how to get your data from the local machine to S3 through python, which might be cleaner."

![Todd_de_Quincey] **@Todd_de_Quincey:** That’ll be it then. Thought I’d check since this connector is so new. Not an issue for prod, as we would be copying from an external S3 stage
Apologies for wasting your time!

![Anna_Geller] **@Anna_Geller:** do you want to open an issue with feature request in the repo?
I'm sure @Andrew_Huang could investigate whether we may add a task for this, this seems to be like a common pattern and quite useful (put is faster than inserts)
never wasting time discussing Prefect Collections! ask us anytime
and btw this comment is from 2019, this might have changed since then

I can open an issue for you here <https://github.com/PrefectHQ/prefect-snowflake/issues/new>
> GitHub

[Todd_de_Quincey]: https://avatars.slack-edge.com/2021-10-31/2681925266081_35f432d9d2c676938ae2_24.png
[Anna_Geller]: https://avatars.slack-edge.com/2022-05-01/3466755334708_dc6796b776c6f927b8f8_24.jpg

### PR DESCRIPTION
**Pull request by peracto**
## Summary

This is a simple change to the existing task to support the Snowflake get/put command.
Snowflake get/put commands don't work when using the Snowflake async_execute method.

NB. This is my first PR - please be kind :)

Apologies ... I'm not sure how to complete the CHANGELOG.md

## Relevant Issue(s)

Add a task to run PUT file commands; closes #8

## Checklist
- [ ] Summarized PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-snowflake/blob/main/CHANGELOG.md)

### CODE REVIEW

**Review by desertaxle — CHANGES_REQUESTED** [review]
Congrats on your first contribution @peracto! 🎉  Thanks for coming up with a solution to this issue!

Instead of adding an option to the current task to control whether the query is executed synchronously or asynchronously, would you be will to create a new task that executes queries synchronously? We could name that task `snowflake_query_sync` and document that it should be used for PUT commands. A separate synchronous task will allow the two query tasks to evolve independently and will be simpler than if we combined them into one.

**Comment by ahuang11** [comment]
I think this closes https://github.com/PrefectHQ/prefect-snowflake/issues/8

**Comment by peracto** [comment]
Yes, I did consider that. I'll update to add the new task.

File: README.md [hunk 1]
```diff
@@ -83,6 +83,29 @@ def snowflake_write_pandas_flow():
         )
 ```
 
+### Execute get/put into Snowflake stage
+
+Snowflake get/put command don't seem to work when executed as an async (Snowflake async as opposed to Python async) job.
+
+For get/put operations use the
```
Comment (ahuang11) [inline README.md:90]:
```suggestion
To execute `get` and `put` statements, use `snowflake_query_sync`.
```

File: README.md [hunk 2]
```diff
@@ -83,6 +83,29 @@ def snowflake_write_pandas_flow():
         )
 ```
 
+### Execute get/put into Snowflake stage
```
Comment (ahuang11) [inline README.md:86]:
```suggestion
### Execute `get` and `put` statements
```

File: README.md [hunk 3]
```diff
@@ -83,6 +83,29 @@ def snowflake_write_pandas_flow():
...
+
+Snowflake get/put command don't seem to work when executed as an async (Snowflake async as opposed to Python async) job.
+
+For get/put operations use the 
+
+```python
+import pandas as pd
```
Comment (ahuang11) [inline README.md:93]:
```suggestion
```

File: prefect_snowflake/database.py [hunk 4]
```diff
@@ -261,3 +261,61 @@ def snowflake_multiquery_flow():
...
+            snowflake_connector = SnowflakeConnector(
+                database="database",
+                warehouse="warehouse",
+                schema="schema",
+                credentials=snowflake_credentials
+            )
+            result = snowflake_query_sync(
+                "put file://afile.csv @mystage;",
+                snowflake_connector,
+                params={"id_param": 1}
```
Comment (ahuang11) [inline prefect_snowflake/database.py:309]:
```suggestion
```

File: prefect_snowflake/database.py [hunk 5]
```diff
@@ -261,3 +261,61 @@ def snowflake_multiquery_flow():
...
+        query: The query to execute against the database.
+        params: The params to replace the placeholders in the query.
+        snowflake_connector: The credentials to use to authenticate.
+        cursor_type: The type of database cursor to use for the query.
+
+    Returns:
+        The output of `response.fetchall()`.
+
+    Examples:
+        Query Snowflake table with the ID value parameterized.
```
Comment (ahuang11) [inline prefect_snowflake/database.py:286]:
```suggestion
        Execute a put statement.
```

File: prefect_snowflake/database.py [hunk 6]
```diff
@@ -261,3 +261,61 @@ def snowflake_multiquery_flow():
...
+        ```python
+        from prefect import flow
+        from prefect_snowflake.credentials import SnowflakeCredentials
+        from prefect_snowflake.database import SnowflakeConnector, snowflake_query
+
+
+        @flow
+        def snowflake_query_flow():
```
Comment (ahuang11) [inline prefect_snowflake/database.py:294]:
```suggestion
        def snowflake_query_sync_flow():
```

**Review by ahuang11 — CHANGES_REQUESTED** [review]
Thanks so much for contributing!

Depending on the built-in error of using execute_async with put/get, I wonder if we should add a check to see if "put" or "get" is used in `execute_query()` and redirect them to use `execute_query_sync` instead. It could also go into another PR.

**Review by desertaxle — APPROVED** [review]
LGTM!

**Review by ahuang11 — APPROVED** [review]
Thanks so much! I'm sure this will be very useful to others too!

**Pull request merged**